Feat: Gateway Upgrade Refactoring + LTS upgrade guide (#6534)
Kong/docs.konghq.com: app/_src/gateway/upgrade/backup-and-restore.md

---
title: Backup and restore
---

Backup and restore methods described in this post only serve as general instructions. You must revise the methods in line with your infrastructure, deployment, and business requirements.

Before upgrading, you should back up data with database tools and optionally declarative tools. We highly recommend doing backup in both methods, which offers you recovery flexibility. The database tools are robust and can restore data instantly compared to the declarative tools. Consequently, in case of data corruption, try to do a database-level restore first, and otherwise bootstrap a new database and then use declarative tools.

> ## Database

Reviewer: Probably a small suggestion, but I think maybe separating the sections by backup and restore vs. database and declarative would work?

## Backup
### database
### declarative
## Restore
### database
### declarative

Just because everyone needs to backup, they might not need to restore, so best to put that at the bottom and separate from backup in case they don't need it?

    specify the admin token. You can specify this token with any decK command:

    ```sh
    deck ping --headers “Kong-Admin-Token: <password>”
    ```

2. Use decK to dump the configuration and store the resulting file in a secure location.
You can back up a particular workspace or all workspaces at once:

    ```sh
    deck dump --all-workspaces -o /path/to/kong_backup.yaml
    deck dump --workspace it_dept -o /path/to/kong_backup.yaml
    ```

    Store the resulting file or files in a safe location.

## Restore Gateway entities

### Database restore

To recover Kong configuration data from a database dump, the operator must make sure the database is correctly prepared. For example, for PostgreSQL:

1. In `kong.conf`, set a database user using the `pg_user` parameter:

    ```
    pg_user = kong
    ```

2. In `kong.conf`, set a database name using the `pg_database` parameter:

    ```
    pg_database = kong
    ```

3. Bootstrap database entities by running `kong migrations bootstrap`.

    Refer to the [`kong migrations` CLI reference](/gateway/{{page.kong_version}}/reference/cli/#kong-migrations) for more information.

> 4. Corresponding to the Postgres database dump, you can now restore the data with the utility `pg_restore`:

Author: I think that whole first section before the comma can go, it doesn't provide any value.

---
title: Backup and restore
content_type: how-to
purpose: Learn how to back up and restore Kong Gateway
---

Before you start any upgrade, back up your {{site.base_gateway}} data. 
Kong supports two back up methods for {{site.base_gateway}} entities: database backup and declarative backup.

We recommend backing up data using both methods, as this offers recovery flexibility.
* The native database tools are robust and can restore data instantly, compared to the declarative tool. 

> * In case of data corruption, try to do a database-level restore first, otherwise bootstrap a new database and use decK to load in entity data.

Reviewer 2: Apart from decK, we have `kong config` command as well. Seems the `kong config` CLI part is removed?

Author: Revisiting this. Previous guidance from Rob (on many tickets that I won't track down now) was that we should not document `kong config` as a viable option _at all_, but based on Rob's response [here](https://docs.google.com/document/d/1ddblT7Nb34LfjGLSoIxKQuN0Q87TKAr8dZffoaID2zw/edit?disco=AAAA3sXNSIY), sounds like we can consider it as an okay secondary option.

* In **traditional and hybrid modes**, decK requires that the database is ready for data export and import. To import or export data using decK, ensure the user and password are initialized, and the database is bootstrapped.
* In **DB-less mode**, decK is the only backup method, as configuration is already managed declaratively.

However, decK has its limitations:

* **Performance**: decK uses the Admin API to read and write entities and might take longer than expected, especially when the number of entities is very large. 

  You can resolve this by increasing the number of threads by passing the flag `--parallelism` to [`deck sync`](/deck/latest/reference/deck_sync/) 
  or [`deck diff`](/deck/latest/reference/deck_diff/), or use decK’s 
  [distributed configuration](/deck/latest/guides/distributed-configuration/) feature.

* **Entities managed by decK**: decK does not manage Enterprise-only entities, like RBAC roles, credentials, keyring, licence, etc. Configure these security related entities separately using Admin API or Kong Manager.
See the reference for [Entities managed by decK](/deck/latest/reference/entities/) for a full list.

> Due to these limitations, we recommend always using the [database dump method](#database-backup) in deployments using a database.

Reviewer 2: ```suggestion
Due to these limitations, we recommend prioritize the [database native method](#database-backup) in deployments using a database.
```

    ```sh
    deck ping
    ```

    If you have RBAC enabled, use the CLI option `--headers` to 
    specify the admin token. You can specify this token with any decK command:

    ```sh
    deck ping --headers “Kong-Admin-Token: <password>”
    ```

2. Use decK to dump the configuration and store the resulting file in a secure location.
You can back up a particular workspace or all workspaces at once:

    ```sh
    deck dump --all-workspaces -o /path/to/kong_backup.yaml
    deck dump --workspace it_dept -o /path/to/kong_backup.yaml
    ```

    Store the resulting file or files in a safe location.

## Restore Gateway entities

### Database restore

To recover {{site.base_gateway}} configuration data from a database dump, the operator must make sure the database is correctly prepared. 
For example, for PostgreSQL:

1. In `kong.conf`, set a database user using the `pg_user` parameter:

    ```
    pg_user = kong
    ```

2. In `kong.conf`, set a database name using the `pg_database` parameter:

    ```
    pg_database = kong
    ```

3. Bootstrap database entities by running `kong migrations bootstrap`.

>     Refer to the [`kong migrations` CLI reference](/gateway/{{page.kong_version}}/reference/cli/#kong-migrations) for more information.

Reviewer 2: Should we suppress the empty line for the two sentences to be in 1 line?

4. You can now restore the data using the utility `pg_restore`:

    ```sh
    pg_restore -U kong -C -d postgres --if-exists --clean kongdb_backup_20230816/
    ```

### Declarative restore

If you need to roll back, change the DB-less {{site.base_gateway}} instance back to the original version, 

> validate the declarative config, then sync it back with your {{site.base_gateway}} instance.

Reviewer 2: decK is only for DB-mode.

1. Check that {{site.base_gateway}} is online:

    ```sh
    deck ping
    ```
2. Validate the declarative config:

    ```sh
    deck validate [--online] -s /path/to/kong_backup.yaml
    ```

3. Once verified, restore a particular workspace or all workspaces at once:

    ```sh
    deck sync --all-workspaces -s /path/to/kong_backup.yaml
    deck sync --workspace it_dept -s /path/to/kong_backup.yaml
    ```

## Keyring materials backup and restore

If you have enabled keyring and data encryption, you must separately back up and restore keyring materials.

{:.important}
> **Caution**: Make sure to store the encryption key in a safe place. 
If the encryption key is lost, you will permanently lose access to the encrypted {{site.base_gateway}} 
configuration data and there is no other way to recover it.

For technical details, refer to the [manual backup method](/gateway/latest/kong-enterprise/db-encryption/#manual-export-and-manual-import) and the [automatic backup method](/gateway/latest/kong-enterprise/db-encryption/#automatic-backup-and-manual-recovery).

> ## kong.conf backup

Reviewer 2: This section contains backup of files other than `kong.conf`. Should we change the title?

---
title: Backup and restore
content_type: how-to
purpose: Learn how to back up and restore Kong Gateway
---

Before you start any upgrade, back up your {{site.base_gateway}} data. 

> Kong supports two back up methods for {{site.base_gateway}} entities: database backup and declarative backup.

Reviewer 2: Please ensure to rename "database backup"  to "database-native backup" on this page. Here is an example:

```suggestion
Kong supports two back up methods for {{site.base_gateway}} entities: [database-native backup](#database-native-backup) and [declarative backup](#declarative-backup).
```

> A database backup backs up the entire {{site.base_gateway}} database, while a declarative backup works by managing declarative configuration files.

Reviewer 2: ```suggestion
A database-native backup backs up the entire {{site.base_gateway}} database, while a declarative backup works by managing declarative configuration files.
```

We recommend backing up data using both methods, as this offers recovery flexibility:

> * The native database tools are robust and can restore data instantly, compared to the declarative tool. 

Reviewer 2: ```suggestion
* The database-native tools are robust and can restore data instantly, compared to the declarative tool. 
```

* In case of data corruption, try to do a database-level restore first, otherwise bootstrap a new database and [declarative tools](#declarative-tools-for-backup-and-restore) to load in entity data.

[Keyring materials](#keyring-materials-backup-and-restore) and {{site.base_gateway}} [configuration files](#other-files) must be backed up separately. 
See their respective sections below for details.

The backup and restore methods described in this guide serve as general instructions.
Revise the methods as necessary to fit your infrastructure, deployment, and business requirements.

## Declarative tools for backup and restore

> Kong ships two declarative backup tools: [decK](/deck/) and the [Kong CLI](/gateway/{{page.kong_version}}/reference/cli/), which support managing {{site.base_gateway}} entities in the declarative format. 

Author: ```suggestion
Kong ships two declarative backup tools: [decK](/deck/) and the [`kong config` CLI](/gateway/{{page.kong_version}}/reference/cli/), which support managing {{site.base_gateway}} entities in the declarative format. 
```

* For **database-backed deployments** (traditional and hybrid mode), backups taken with either of these tools serve as an extra safeguard layer. If the database backup or restore corrupts the database, you can fall back to declarative files for restoring data.

    Both tools require the database to be ready for data export and import. To import or export data using these tools, ensure the user and password are initialized, and the database is bootstrapped.

> * For **DB-less deployments**, the Kong CLI is the only backup and restore tool available. decK is not supported, as it requires the Admin API to function.

Reviewer 2: ```suggestion
* For **DB-less deployments**, neither tools work Please backup the your declarative files manually.
```

> decK is generally more powerful than the Kong CLI. It has more features, invalidates the cache automatically, and fetches entities from the database instead of the LRU cache. Additionally, it overwrites entities instead of patching, so that the database has the exact copy of the config that you provide.

Author: ```suggestion
decK is generally more powerful than the `kong config` CLI. It has more features, invalidates the cache automatically, and fetches entities from the database instead of the LRU cache. Additionally, it overwrites entities instead of patching, so that the database has the exact copy of the config that you provide.
```

However, decK also has its limitations:

> * **Availability**: decK requires {{site.base_gateway}} to be online, while the Kong CLI doesn't. 

Author: ```suggestion
* **Availability**: decK requires {{site.base_gateway}} to be online, while the `kong config` CLI doesn't. 
```

* **Performance**: decK uses the Admin API to read and write entities and might take longer than expected, especially when the number of entities is very large. 

  You can resolve this by increasing the number of threads by passing the flag `--parallelism` to [`deck sync`](/deck/latest/reference/deck_sync/) 
  or [`deck diff`](/deck/latest/reference/deck_diff/), or use decK’s 
  [distributed configuration](/deck/latest/guides/distributed-configuration/) feature.

* **Entities managed by decK**: decK does not manage Enterprise-only entities, like RBAC roles, credentials, keyring, licence, etc. Configure these security related entities separately using Admin API or Kong Manager.
See the reference for [Entities managed by decK](/deck/latest/reference/entities/) for a full list.

> Due to these limitations, we recommend prioritizing the [database native method](#database-backup) in deployments using a database.

Reviewer 2: ```suggestion
Due to these limitations, we recommend prioritizing the [database-native method](#database-native-backup) in deployments using a database.
```

## Back up Gateway entities

### Database-native backup

> When upgrading your {{site.base_gateway}} to a newer version, you have to perform a database migration using the `kong migrations` utility. The `kong migrations` commands are not reversible. We recommend backing up data before any starting any upgrade in case of any migration issues.

Reviewer 2: Add a ref link to `kong migrations` CLI.

If you are running {{site.base_gateway}} with a database, run a database dump of raw data so that you can recover the database quickly in a database-native way. This is the recommended way to back up {{site.base_gateway}}.

> With PostgreSQL, you can dump data in text format, tar format (no compression), or directory format (with compression) using the utility `pg_dump`. For example:

Reviewer 2: ```suggestion
With PostgreSQL, you can dump data in *text* format, *tar* format (no compression), or *directory* format (with compression) using the utility `pg_dump`. For example:
```

```sh
pg_dump -U kong -d kong -F d -f kongdb_backup_20230816
```

Use the CLI option `-d` to specify the database (for example, `kong`) to export, especially when the PostgreSQL instance also serves applications other than {{site.base_gateway}}.

### Declarative backup

> {% navtabs %}

Reviewer 2: Great style!

    deck validate [--online] -s /path/to/kong_backup.yaml
    ```

3. Once verified, restore a particular workspace or all workspaces at once:

    ```sh
    deck sync --all-workspaces -s /path/to/kong_backup.yaml
    ```
    or

    ```sh
    deck sync --workspace it_dept -s /path/to/kong_backup.yaml
    ```

{% endnavtab %}
{% navtab Traditional or hybrid mode - Kong CLI %}

If you backed up {{site.base_gateway}} database using `kong config db_export`, 
use the Kong CLI to restore your configuration from the backup declarative config file.

1. Validate the backup configuration file before restoring it:

    ```sh
    kong config parse /path/to/kong_backup.yaml
    ```

2. Import entities into your database:

    ```sh
    kong config db_import /path/to/kong_backup.yaml
    ```

2. Restart or reload your {{site.base_gateway}} instance:

    ```sh
    kong restart
    ```

    or

    ```sh
    kong reload
    ```

{% endnavtab %}
{% navtab DB-less mode %}

In DB-less mode, use the Kong CLI to restore your configuration from a declarative config file.

1. Validate the backup configuration file before restoring it:

    ```sh
    kong config parse /path/to/kong_backup.yaml
    ```

> 2. Restart or reload your {{site.base_gateway}} instance using the backup configuration file:
> 
>     ```sh
>     kong restart -c /path/to/kong_backup.yaml
>     ```
> 
>     or
> 
>     ```sh
>     kong reload -c /path/to/kong_backup.yaml
>     ```

Reviewer 2: This is NOT right! To restore with the declarative backup file, please point [declarative_config](https://docs.konghq.com/gateway/latest/reference/configuration/#declarative_config) in `kong.conf` to `/path/to/kong_backup.yaml`.

```suggestion
2. Restart or reload your {{site.base_gateway}} instance using the backup configuration file:

    ```sh
    export KONG_DECLARATIVE_CONFIG=kong.yml; kong restart -c /path/to/kong_backup.yaml
    # -or-
    export KONG_DECLARATIVE_CONFIG=kong.yml; kong reload -c /path/to/kong_backup.yaml
    ```

    Alternatively, post the declarative backup file to `:8001/config` endpoint.

    ```sh
    curl -sS http://localhost:8001/config?check_hash=1 -F 'config=@kong_backup.yaml' ; echo
    ```
```
